L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc

```diff
@@ -0,0 +1,136 @@
+                                                                                                  nBinsEta(conf.getParameter<int>("nBinsEta")),
+                                                                                                  nBinsPhi(conf.getParameter<int>("nBinsPhi")),
+                                                                                                  binsEta(conf.getParameter<std::vector<double> >("binsEta")),
+                                                                                                  binsPhi(conf.getParameter<std::vector<double> >("binsPhi")) {
+
+
+  if(!binsEta.empty() && ((unsigned int)(binsEta.size()) != nBinsEta+1)) {
+    throw edm::Exception(edm::errors::Configuration, "Configuration")
+      << "HGCalTriggerTowerGeometryHelper nBinsEta for the tower map not consistent with binsEta size"<<std::endl;
+  }
+
+  if(!binsPhi.empty() && ((unsigned int)(binsPhi.size()) != nBinsPhi+1)) {
+    throw edm::Exception(edm::errors::Configuration, "Configuration")
+      << "HGCalTriggerTowerGeometryHelper nBinsPhi for the tower map not consistent with binsPhi size"<<std::endl;
+  }
+
+  // if the bin vecctor is empty we assume the bins to be regularly spaced
+  if(binsEta.empty()) {
+    for(unsigned int bin1 = 0; bin1 != nBinsEta+1; bin1++) {
+      binsEta.push_back(minEta+bin1*((maxEta-minEta)/nBinsEta));
+    }
+  }
+
+  // if the bin vecctor is empty we assume the bins to be regularly spaced
+  if(binsPhi.empty()) {
+    for(unsigned int bin2 = 0; bin2 != nBinsPhi+1; bin2++) {
+      binsPhi.push_back(minPhi+bin2*((maxPhi-minPhi)/nBinsPhi));
+    }
+  }
+
+  for(int zside = -1; zside <= 1; zside+=2) {
+    for(unsigned int bin1 = 0; bin1 != nBinsEta; bin1++) {
+      for(unsigned int bin2 = 0; bin2 != nBinsPhi; bin2++) {
+        l1t::HGCalTowerID towerId(zside, bin1, bin2);
+        tower_coords_.emplace_back(towerId.rawId(),
+                                   (binsEta[bin1+1] + binsEta[bin1])/2,
+                                   (binsPhi[bin2+1] + binsPhi[bin2])/2);
+      }
+    }
+  }
+
+  if(conf.getParameter<bool>("readMappingFile")) {
+    // We read the TC to TT mapping from file,
+    // otherwise we derive the TC to TT mapping on the fly from eta-phi coord. of the TCs
+    std::ifstream l1tTriggerTowerMappingStream(conf.getParameter<edm::FileInPath>("L1TTriggerTowerMapping").fullPath());
+    if(!l1tTriggerTowerMappingStream.is_open()) {
+        throw cms::Exception("MissingDataFile")
+            << "Cannot open HGCalTriggerGeometry L1TTriggerTowerMapping file\n";
+    }
+
+    unsigned trigger_cell_id = 0;
+    unsigned short ix = 0;
+    unsigned short iy = 0;
+
+    for(; l1tTriggerTowerMappingStream >> trigger_cell_id >> ix >> iy;) {
+      HGCalDetId detId(trigger_cell_id);
+      int zside = detId.zside();
+      l1t::HGCalTowerID towerId(zside, ix, iy);
+      cells_to_trigger_towers_[trigger_cell_id] = towerId.rawId();
```

Comment:
Indeed there is no enforcement at the moment. Once could look-up for the coordinate of each new ID while filling the mapping. This is not very efficient (given the size of the map) but if you feel it safer I can code it.

Reply:
One could fill two (ordered) sets of tower ids corresponding to `cells_to_trigger_towers_` and to `tower_coords_` and check that they are equal. There is not so many towers (compared to trigger cells) so it shouldn't be too memory consuming nor too long.

Reply:
Actually it would be enough to check that the iEta and iPhi read from the mapping file are < of `nBinsEta_` and `nBinsPhi_` respectively.
Also because it is possible, in some geometry that some of towers are defined but actually not containing any TC (for example if one has fine binning towards the bunday of the detector).
Would that be an acceptable test?

Reply:
Ok. Yes, it should be safe enough.